cpp/src/parquet/encryption/encryption.cc

```diff
@@ -182,79 +175,75 @@ FileEncryptionProperties::Builder::disable_aad_prefix_storage() {
 }
 
 ColumnEncryptionProperties::ColumnEncryptionProperties(bool encrypted,
-                                                       const std::string& column_path,
-                                                       const std::string& key,
-                                                       const std::string& key_metadata)
-    : column_path_(column_path) {
-  DCHECK(!column_path.empty());
+                                                       std::string column_path,
+                                                       encryption::SecureString key,
+                                                       std::string key_metadata)
+    : column_path_(std::move(column_path)),
```

Comment:
This moves member initialization from the body, which is the more standard way. However, the body could accidentally check input arguments, which after `std::move` are "empty".

While this is correct:
``
      encrypted_with_footer_key_(encrypted && key.empty()),
      key_(std::move(key)),
```
if members are shuffled around, then this becomes incorrect:
```
      key_(std::move(key)),
      encrypted_with_footer_key_(encrypted && key.empty()),
```
as `encrypted_with_footer_key_` is initialized on the "empty" `key`.

Do you consider this pattern too risky for future bugs? Happy to revert this initialization.

Reply:
Member initializer order is asserted to match member declaration order in a lint pass. (Member initialization proceeds in declaration order regardless of member initializer order.) If the member declarations and initializers were both reordered, then
```c++
key_(std::move(key)),
encrypted_with_footer_key_(encrypted && key.empty()),
```
is use-after-move which should fail a separate lint check. Furthermore, that could be fixed by writing
```c++
key_(std::move(key)),
encrypted_with_footer_key_(encrypted && key_.empty()),
```

So I don't think it's necessary to revert.